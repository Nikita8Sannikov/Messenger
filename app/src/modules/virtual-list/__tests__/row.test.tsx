import { render, screen } from "@testing-library/react";

import ListRow from "@modules/virtual-list/row";

describe("ListRow", () => {
	const mockStyle = { top: 0, height: 100 };
	const mockOptions = {
		cols: 3,
		gap: 10,
		itemHeight: 200,
	};

	const mockRenderCb = ({ item, itemWidth, itemHeight }) => (
		<div data-testid={`item-${item.id}`}>
			{item.name} ({itemWidth}x{itemHeight})
		</div>
	);

	it("renders loading row when no items and hasMore is true", () => {
		const data = {
			items: [],
			itemWidth: 300,
			hasMore: true,
		};

		const { container } = render(
			<ListRow
				index={0}
				style={mockStyle}
				data={data}
				options={mockOptions}
				renderCb={mockRenderCb}
			/>,
		);

		expect(screen.getByRole("progressbar")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("renders items row correctly", () => {
		const data = {
			items: [
				{ id: "1", name: "Item 1" },
				{ id: "2", name: "Item 2" },
				{ id: "3", name: "Item 3" },
			],
			itemWidth: 300,
			hasMore: false,
		};

		const { container } = render(
			<ListRow
				index={0}
				style={mockStyle}
				data={data}
				options={mockOptions}
				renderCb={mockRenderCb}
			/>,
		);

		expect(screen.getByTestId("item-1")).toBeInTheDocument();
		expect(screen.getByTestId("item-2")).toBeInTheDocument();
		expect(screen.getByTestId("item-3")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("renders correct number of items based on cols", () => {
		const data = {
			items: [
				{ id: "1", name: "Item 1" },
				{ id: "2", name: "Item 2" },
				{ id: "3", name: "Item 3" },
				{ id: "4", name: "Item 4" },
			],
			itemWidth: 300,
			hasMore: false,
		};

		const { container } = render(
			<ListRow
				index={0}
				style={mockStyle}
				data={data}
				options={{ ...mockOptions, cols: 2 }}
				renderCb={mockRenderCb}
			/>,
		);

		expect(screen.getByTestId("item-1")).toBeInTheDocument();
		expect(screen.getByTestId("item-2")).toBeInTheDocument();
		expect(screen.queryByTestId("item-3")).not.toBeInTheDocument();
		expect(screen.queryByTestId("item-4")).not.toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("applies correct styles to row", () => {
		const data = {
			items: [ { id: "1", name: "Item 1" } ],
			itemWidth: 300,
			hasMore: false,
		};

		const { container } = render(
			<ListRow
				index={0}
				style={mockStyle}
				data={data}
				options={mockOptions}
				renderCb={mockRenderCb}
			/>,
		);

		const row = container.firstChild as HTMLElement;
		expect(row).toHaveStyle({
			top: "0px",
			height: "100px",
			gap: "10px",
		});
		expect(container).toMatchSnapshot();
	});
});
