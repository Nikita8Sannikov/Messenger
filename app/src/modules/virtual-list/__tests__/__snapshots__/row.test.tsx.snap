// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListRow applies correct styles to row 1`] = `
<div>
  <div
    class="list__row"
    style="top: 0px; height: 100px; gap: 10px;"
  >
    <div>
      <div
        data-testid="item-1"
      >
        Item 1
         (
        300
        x
        200
        )
      </div>
    </div>
  </div>
</div>
`;

exports[`ListRow renders correct number of items based on cols 1`] = `
<div>
  <div
    class="list__row"
    style="top: 0px; height: 100px; gap: 10px;"
  >
    <div>
      <div
        data-testid="item-1"
      >
        Item 1
         (
        300
        x
        200
        )
      </div>
    </div>
    <div>
      <div
        data-testid="item-2"
      >
        Item 2
         (
        300
        x
        200
        )
      </div>
    </div>
  </div>
</div>
`;

exports[`ListRow renders items row correctly 1`] = `
<div>
  <div
    class="list__row"
    style="top: 0px; height: 100px; gap: 10px;"
  >
    <div>
      <div
        data-testid="item-1"
      >
        Item 1
         (
        300
        x
        200
        )
      </div>
    </div>
    <div>
      <div
        data-testid="item-2"
      >
        Item 2
         (
        300
        x
        200
        )
      </div>
    </div>
    <div>
      <div
        data-testid="item-3"
      >
        Item 3
         (
        300
        x
        200
        )
      </div>
    </div>
  </div>
</div>
`;

exports[`ListRow renders loading row when no items and hasMore is true 1`] = `
<div>
  <div
    class="list__loading-row"
    style="top: 0px; height: 100px;"
  >
    <div
      class="spinner"
      data-testid="spinner"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
